hal/spi_api.h

```diff
@@ -19,220 +18,346 @@
 #ifndef MBED_SPI_API_H
 #define MBED_SPI_API_H
 
+#include <stdbool.h>
+#include <stdint.h>
 #include "device.h"
-#include "hal/dma_api.h"
-#include "hal/buffer.h"
 
 #if DEVICE_SPI
+/**
+ * \defgroup hal_spi SPI: Serial peripheral interface HAL API.
+ * Low level interface to the serial peripheral interface of a target.
+ *
+ * ## Programming model
+ * This is an OOP like API. spi_t is used a "this".<br/>
+ * Methods not having this type as their first parameter are **class** methods.<br/>
+ * Methods with an spi_t object as their first parameter are **instance** methods.<br/>
+ * `spi_init()` acts as the "constructor"(/initialiser) of this *class*.<br/>
+ * `spi_free()` acts as the "destructor"(/deinitialiser) of this *class*.<br/>
+ *
+ * ## General requirements
+ * - Multiple instances can share pins such as MISO/MOSI/MCLK but not their SS pin. This pin is used
+ *   to distinguish communication channels (master or slaves).
+ * - As the instance allocation is at the user's responsibility, the hal implementation cannot keep
+ *   a copy of the pointer (keep an alias to the object) as it may be stack allocated and/or moved.
+ * - All method must be thread safe at HAL level.
+ * - The hal implementation must support both clock phases.
+ * - The hal implementation must support both clock polarities.
+ * - The hal implementation must support either master or slave or both modes.
+ * - The hal implementation must support either lsb first or msb first or both bit ordering.
+ * - The hal implementation must support either continuous or noncontinuous or both modes.
+ * - The hal implementation must support at least 1 word length in the range 4 to 32.
+ * - The hal implementation must support the simplex and full duplex modes.
+ * - The hal implementation may or may not support the half-duplex, dual and/or quad-io modes.
+ * - The accuracy of the clock generated in master mode by the hal implementation must be within +/-
+ *   0.1% of the user requested frequency.
+ *
+ * # Defined behaviours
+ * - spi_get_capabilities(..)
+ *   - returns NULL if passed a NULL pointer.
+ *   - returns NULL if the peripheral cannot be identified.
+ *   - returns a constant pointer to a constant structure describing the capabilities of the
+ *     related peripheral.
+ *   - returned value is independent from the SPI channel usage.
+ * - spi_init(..)
+ *   - returns SPI_RESULT_INVALID_PARAM if any of the parameters is NULL.
```

Comment:
Defining the behavior of how to handle invalid parameters will add boilerplate code to every implementation. What are your thoughts on moving this to the undefined behavior section and enforcing valid params at the driver layer?

Reply:
I usually have this checks done at API boundaries (such as HAL APIs).
IIRC Misra C recommend that to be done in every function.
I like to think of checking only at API boundaries as a "trade off" 😮

Reply:
I am still not convinced about this. It's error prone and will be copy-paste everywhere. One wrong check and we will be fixing hundreds implementations. I rather would go as @c1728p9 suggested. HAL does not check this and upper layer must.

Reply:
If your concerns are only about validating "non-NULL pointer" we could provide a function similar to what we have on the sai feature branch : https://github.com/ARMmbed/mbed-os/blob/feature-hal-spec-sai/hal/mbed_sai_api.c#L95

However, if you have doubt about any input validations :
These APIs are usually implemented at "device family" layer which drastically reduce the number of implementations to a few per silicon vendor (~2-4).
There is currently `55` file implementing `spi_init`  for `15` vendors.
`13` from NXP where most of them only differ by the pins being declared in the `*.c` and few other thing that could easily be `#if` gated.
5 out of the 10 Freescale's `MCUXpresso` are exactly the same file, and 2 differ because they implement the Async API as well.

Some good code factorisation would be made while implementing this new API thus reducing the number of place where an error could be made.

Reply:
> If your concerns are only about validating "non-NULL pointer" we could provide a function similar to what we have on the sai feature branch : https://github.com/ARMmbed/mbed-os/blob/feature-hal-spec-sai/hal/mbed_sai_api.c#L95

I think this is it, @c1728p9 was also asking for (although the question was move it upper).